File: Assets/Script/Lib/UIX/paint/primitives.lua
-- [ts]: primitives.ts
local ____exports = {} -- 1
local ____Dora = require("Dora") -- 1
local Color = ____Dora.Color -- 1
local nvg = require("nvg") -- 2
local ____types = require("UIX.types") -- 5
local clamp = ____types.clamp -- 5
local ____color = require("UIX.paint.color") -- 6
local withAlpha = ____color.withAlpha -- 6
function ____exports.rect(width, height) -- 8
	return {x = 0, y = 0, width = width, height = height} -- 9
end -- 8
function ____exports.roundedPanel(ctx, r, options) -- 12
	local theme = ctx.theme -- 17
	local radius = options and options.radius or theme.radius.md -- 18
	local elevated = (options and options.elevated) ~= false -- 19
	if elevated then -- 19
		nvg.BeginPath() -- 21
		nvg.RoundedRect( -- 22
			r.x + 2, -- 22
			r.y + 3, -- 22
			r.width, -- 22
			r.height, -- 22
			radius -- 22
		) -- 22
		nvg.FillColor(Color(withAlpha(4278190080, theme.painter.shadowAlpha * ctx.opacity))) -- 23
		nvg.Fill() -- 24
	end -- 24
	local fill = (options and options.variant) == "solid" and theme.colors.surface.base or theme.colors.surface.raised -- 26
	nvg.BeginPath() -- 27
	nvg.RoundedRect( -- 28
		r.x, -- 28
		r.y, -- 28
		r.width, -- 28
		r.height, -- 28
		radius -- 28
	) -- 28
	nvg.FillColor(Color(withAlpha(fill, ctx.opacity))) -- 29
	nvg.Fill() -- 30
	nvg.BeginPath() -- 31
	nvg.RoundedRect( -- 32
		r.x + 0.5, -- 32
		r.y + 0.5, -- 32
		r.width - 1, -- 32
		r.height - 1, -- 32
		radius -- 32
	) -- 32
	nvg.StrokeWidth(theme.stroke.hairline) -- 33
	nvg.StrokeColor(Color(withAlpha(theme.colors.line.normal, ctx.opacity))) -- 34
	nvg.Stroke() -- 35
end -- 12
function ____exports.buttonSurface(ctx, r, options) -- 38
	local theme = ctx.theme -- 42
	local state = ctx.state -- 43
	local variant = options and options.variant or "primary" -- 44
	local radius = options and options.radius or theme.radius.md -- 45
	local fill = theme.colors.surface.raised -- 46
	local stroke = theme.colors.line.normal -- 47
	if variant == "primary" then -- 47
		stroke = theme.colors.accent.primary -- 48
	end -- 48
	if variant == "secondary" then -- 48
		stroke = theme.colors.accent.secondary -- 49
	end -- 49
	if variant == "danger" then -- 49
		stroke = theme.colors.state.danger -- 50
	end -- 50
	if variant == "glass" then -- 50
		fill = withAlpha(theme.colors.surface.raised, 0.78) -- 51
	end -- 51
	if variant == "ghost" then -- 51
		fill = withAlpha(theme.colors.surface.raised, (state.hovered or state.pressed) and 0.45 or 0.08) -- 52
	end -- 52
	if state.selected then -- 52
		fill = withAlpha(theme.colors.accent.primary, 0.35) -- 53
	end -- 53
	if state.pressed then -- 53
		fill = theme.colors.surface.sunken -- 54
	end -- 54
	if state.disabled then -- 54
		fill = withAlpha(theme.colors.surface.sunken, theme.painter.disabledAlpha) -- 56
		stroke = theme.colors.line.subtle -- 57
	end -- 57
	nvg.BeginPath() -- 59
	nvg.RoundedRect( -- 60
		r.x, -- 60
		r.y, -- 60
		r.width, -- 60
		r.height, -- 60
		radius -- 60
	) -- 60
	nvg.FillColor(Color(withAlpha(fill, ctx.opacity))) -- 61
	nvg.Fill() -- 62
	if not state.disabled and (state.hovered or state.selected) then -- 62
		nvg.BeginPath() -- 64
		nvg.RoundedRect( -- 65
			r.x + 1, -- 65
			r.y + 1, -- 65
			r.width - 2, -- 65
			r.height - 2, -- 65
			radius -- 65
		) -- 65
		nvg.FillColor(Color(withAlpha(theme.colors.accent.primary, 0.12 * ctx.opacity))) -- 66
		nvg.Fill() -- 67
	end -- 67
	nvg.BeginPath() -- 69
	nvg.RoundedRect( -- 70
		r.x + 0.5, -- 70
		r.y + 0.5, -- 70
		r.width - 1, -- 70
		r.height - 1, -- 70
		radius -- 70
	) -- 70
	nvg.StrokeWidth(state.focused and theme.stroke.normal or theme.stroke.hairline) -- 71
	nvg.StrokeColor(Color(withAlpha(stroke, state.disabled and 0.45 or ctx.opacity))) -- 72
	nvg.Stroke() -- 73
end -- 38
function ____exports.progressTrack(ctx, r) -- 76
	local theme = ctx.theme -- 77
	nvg.BeginPath() -- 78
	nvg.RoundedRect( -- 79
		r.x, -- 79
		r.y, -- 79
		r.width, -- 79
		r.height, -- 79
		math.min(theme.radius.sm, r.height / 2) -- 79
	) -- 79
	nvg.FillColor(Color(withAlpha(theme.colors.surface.sunken, ctx.opacity))) -- 80
	nvg.Fill() -- 81
	nvg.StrokeWidth(theme.stroke.hairline) -- 82
	nvg.StrokeColor(Color(withAlpha(theme.colors.line.subtle, ctx.opacity))) -- 83
	nvg.Stroke() -- 84
end -- 76
function ____exports.progressFill(ctx, r, progress, variant) -- 87
	local theme = ctx.theme -- 88
	local p = clamp(progress, 0, 1) -- 89
	if p <= 0 then -- 89
		return -- 90
	end -- 90
	local color = theme.colors.accent.primary -- 91
	if variant == "health" then -- 91
		color = theme.colors.state.danger -- 92
	end -- 92
	if variant == "mana" then -- 92
		color = theme.colors.state.mana -- 93
	end -- 93
	if variant == "shield" then -- 93
		color = theme.colors.state.shield -- 94
	end -- 94
	if variant == "warm" then -- 94
		color = theme.colors.accent.warm -- 95
	end -- 95
	local width = math.max(1, r.width * p) -- 96
	nvg.BeginPath() -- 97
	nvg.RoundedRect( -- 98
		r.x, -- 98
		r.y, -- 98
		width, -- 98
		r.height, -- 98
		math.min(theme.radius.sm, r.height / 2) -- 98
	) -- 98
	nvg.FillColor(Color(withAlpha(color, ctx.opacity))) -- 99
	nvg.Fill() -- 100
end -- 87
function ____exports.focusRing(ctx, r, options) -- 103
	if not ctx.state.focused or ctx.state.disabled then -- 103
		return -- 108
	end -- 108
	local theme = ctx.theme -- 109
	local inset = options and options.inset or -2 -- 110
	local radius = options and options.radius or theme.radius.md + 2 -- 111
	nvg.BeginPath() -- 112
	nvg.RoundedRect( -- 113
		r.x + inset, -- 113
		r.y + inset, -- 113
		r.width - inset * 2, -- 113
		r.height - inset * 2, -- 113
		radius -- 113
	) -- 113
	nvg.StrokeWidth(theme.stroke.focus) -- 114
	nvg.StrokeColor(Color(withAlpha(options and options.color or theme.colors.focus.ring, ctx.opacity))) -- 115
	nvg.Stroke() -- 116
end -- 103
function ____exports.cooldownMask(ctx, r, progress) -- 119
	local p = clamp(progress, 0, 1) -- 120
	if p <= 0 then -- 120
		return -- 121
	end -- 121
	local height = r.height * p -- 122
	nvg.BeginPath() -- 123
	nvg.RoundedRect( -- 124
		r.x, -- 124
		r.y, -- 124
		r.width, -- 124
		height, -- 124
		ctx.theme.radius.md -- 124
	) -- 124
	nvg.FillColor(Color(withAlpha(ctx.theme.colors.accent.primary, 0.34 * ctx.opacity))) -- 125
	nvg.Fill() -- 126
end -- 119
local function itemQualityColor(ctx, quality) -- 129
	if quality == "rare" then -- 129
		return ctx.theme.colors.accent.secondary -- 130
	end -- 130
	if quality == "epic" then -- 130
		return 4290272511 -- 131
	end -- 131
	if quality == "legendary" then -- 131
		return ctx.theme.colors.accent.warm -- 132
	end -- 132
	if quality == "common" then -- 132
		return ctx.theme.colors.line.strong -- 133
	end -- 133
	return ctx.theme.colors.line.subtle -- 134
end -- 129
function ____exports.itemSlotSurface(ctx, r, quality, selected) -- 137
	local theme = ctx.theme -- 138
	local radius = theme.radius.md -- 139
	local state = ctx.state -- 140
	local fill = state.disabled and withAlpha(theme.colors.surface.sunken, theme.painter.disabledAlpha) or theme.colors.surface.sunken -- 141
	local stroke = itemQualityColor(ctx, quality) -- 142
	if state.pressed then -- 142
		stroke = theme.colors.text.primary -- 143
	end -- 143
	if selected == true or state.selected then -- 143
		stroke = theme.colors.accent.primary -- 144
	end -- 144
	nvg.BeginPath() -- 145
	nvg.RoundedRect( -- 146
		r.x, -- 146
		r.y, -- 146
		r.width, -- 146
		r.height, -- 146
		radius -- 146
	) -- 146
	nvg.FillColor(Color(withAlpha(fill, ctx.opacity))) -- 147
	nvg.Fill() -- 148
	if quality ~= "empty" and not state.disabled then -- 148
		nvg.BeginPath() -- 150
		nvg.RoundedRect( -- 151
			r.x + 3, -- 151
			r.y + 3, -- 151
			r.width - 6, -- 151
			r.height - 6, -- 151
			math.max(1, radius - 2) -- 151
		) -- 151
		nvg.FillColor(Color(withAlpha(stroke, (selected == true or state.selected) and 0.24 or 0.12))) -- 152
		nvg.Fill() -- 153
	end -- 153
	nvg.BeginPath() -- 155
	nvg.RoundedRect( -- 156
		r.x + 0.5, -- 156
		r.y + 0.5, -- 156
		r.width - 1, -- 156
		r.height - 1, -- 156
		radius -- 156
	) -- 156
	nvg.StrokeWidth((selected == true or state.selected) and theme.stroke.normal or theme.stroke.hairline) -- 157
	nvg.StrokeColor(Color(withAlpha(stroke, state.disabled and 0.38 or ctx.opacity))) -- 158
	nvg.Stroke() -- 159
end -- 137
return ____exports -- 137

File: Assets/Script/Lib/UIX/paint/primitives.ts
import { Color } from "Dora";
import * as nvg from "nvg";
import type { PaintContext } from "UIX/paint/PaintNode";
import type { ProgressVariant, Rect, UiVariant } from "UIX/types";
import { clamp } from "UIX/types";
import { withAlpha } from "UIX/paint/color";

export function rect(this: void, width: number, height: number): Rect {
	return { x: 0, y: 0, width, height };
}

export function roundedPanel(this: void, ctx: PaintContext, r: Rect, options?: {
	variant?: "default" | "glass" | "solid";
	radius?: number;
	elevated?: boolean;
}): void {
	const theme = ctx.theme;
	const radius = options?.radius ?? theme.radius.md;
	const elevated = options?.elevated !== false;
	if (elevated) {
		nvg.BeginPath();
		nvg.RoundedRect(r.x + 2, r.y + 3, r.width, r.height, radius);
		nvg.FillColor(Color(withAlpha(0xff000000, theme.painter.shadowAlpha * ctx.opacity)));
		nvg.Fill();
	}
	const fill = options?.variant === "solid" ? theme.colors.surface.base : theme.colors.surface.raised;
	nvg.BeginPath();
	nvg.RoundedRect(r.x, r.y, r.width, r.height, radius);
	nvg.FillColor(Color(withAlpha(fill, ctx.opacity)));
	nvg.Fill();
	nvg.BeginPath();
	nvg.RoundedRect(r.x + 0.5, r.y + 0.5, r.width - 1, r.height - 1, radius);
	nvg.StrokeWidth(theme.stroke.hairline);
	nvg.StrokeColor(Color(withAlpha(theme.colors.line.normal, ctx.opacity)));
	nvg.Stroke();
}

export function buttonSurface(this: void, ctx: PaintContext, r: Rect, options?: {
	variant?: UiVariant;
	radius?: number;
}): void {
	const theme = ctx.theme;
	const state = ctx.state;
	const variant = options?.variant ?? "primary";
	const radius = options?.radius ?? theme.radius.md;
	let fill = theme.colors.surface.raised;
	let stroke = theme.colors.line.normal;
	if (variant === "primary") stroke = theme.colors.accent.primary;
	if (variant === "secondary") stroke = theme.colors.accent.secondary;
	if (variant === "danger") stroke = theme.colors.state.danger;
	if (variant === "glass") fill = withAlpha(theme.colors.surface.raised, 0.78);
	if (variant === "ghost") fill = withAlpha(theme.colors.surface.raised, state.hovered || state.pressed ? 0.45 : 0.08);
	if (state.selected) fill = withAlpha(theme.colors.accent.primary, 0.35);
	if (state.pressed) fill = theme.colors.surface.sunken;
	if (state.disabled) {
		fill = withAlpha(theme.colors.surface.sunken, theme.painter.disabledAlpha);
		stroke = theme.colors.line.subtle;
	}
	nvg.BeginPath();
	nvg.RoundedRect(r.x, r.y, r.width, r.height, radius);
	nvg.FillColor(Color(withAlpha(fill, ctx.opacity)));
	nvg.Fill();
	if (!state.disabled && (state.hovered || state.selected)) {
		nvg.BeginPath();
		nvg.RoundedRect(r.x + 1, r.y + 1, r.width - 2, r.height - 2, radius);
		nvg.FillColor(Color(withAlpha(theme.colors.accent.primary, 0.12 * ctx.opacity)));
		nvg.Fill();
	}
	nvg.BeginPath();
	nvg.RoundedRect(r.x + 0.5, r.y + 0.5, r.width - 1, r.height - 1, radius);
	nvg.StrokeWidth(state.focused ? theme.stroke.normal : theme.stroke.hairline);
	nvg.StrokeColor(Color(withAlpha(stroke, state.disabled ? 0.45 : ctx.opacity)));
	nvg.Stroke();
}

export function progressTrack(this: void, ctx: PaintContext, r: Rect): void {
	const theme = ctx.theme;
	nvg.BeginPath();
	nvg.RoundedRect(r.x, r.y, r.width, r.height, math.min(theme.radius.sm, r.height / 2));
	nvg.FillColor(Color(withAlpha(theme.colors.surface.sunken, ctx.opacity)));
	nvg.Fill();
	nvg.StrokeWidth(theme.stroke.hairline);
	nvg.StrokeColor(Color(withAlpha(theme.colors.line.subtle, ctx.opacity)));
	nvg.Stroke();
}

export function progressFill(this: void, ctx: PaintContext, r: Rect, progress: number, variant: ProgressVariant): void {
	const theme = ctx.theme;
	const p = clamp(progress, 0, 1);
	if (p <= 0) return;
	let color = theme.colors.accent.primary;
	if (variant === "health") color = theme.colors.state.danger;
	if (variant === "mana") color = theme.colors.state.mana;
	if (variant === "shield") color = theme.colors.state.shield;
	if (variant === "warm") color = theme.colors.accent.warm;
	const width = math.max(1, r.width * p);
	nvg.BeginPath();
	nvg.RoundedRect(r.x, r.y, width, r.height, math.min(theme.radius.sm, r.height / 2));
	nvg.FillColor(Color(withAlpha(color, ctx.opacity)));
	nvg.Fill();
}

export function focusRing(this: void, ctx: PaintContext, r: Rect, options?: {
	inset?: number;
	radius?: number;
	color?: number;
}): void {
	if (!ctx.state.focused || ctx.state.disabled) return;
	const theme = ctx.theme;
	const inset = options?.inset ?? -2;
	const radius = options?.radius ?? theme.radius.md + 2;
	nvg.BeginPath();
	nvg.RoundedRect(r.x + inset, r.y + inset, r.width - inset * 2, r.height - inset * 2, radius);
	nvg.StrokeWidth(theme.stroke.focus);
	nvg.StrokeColor(Color(withAlpha(options?.color ?? theme.colors.focus.ring, ctx.opacity)));
	nvg.Stroke();
}

export function cooldownMask(this: void, ctx: PaintContext, r: Rect, progress: number): void {
	const p = clamp(progress, 0, 1);
	if (p <= 0) return;
	const height = r.height * p;
	nvg.BeginPath();
	nvg.RoundedRect(r.x, r.y, r.width, height, ctx.theme.radius.md);
	nvg.FillColor(Color(withAlpha(ctx.theme.colors.accent.primary, 0.34 * ctx.opacity)));
	nvg.Fill();
}

function itemQualityColor(this: void, ctx: PaintContext, quality: string): number {
	if (quality === "rare") return ctx.theme.colors.accent.secondary;
	if (quality === "epic") return 0xffb85cff;
	if (quality === "legendary") return ctx.theme.colors.accent.warm;
	if (quality === "common") return ctx.theme.colors.line.strong;
	return ctx.theme.colors.line.subtle;
}

export function itemSlotSurface(this: void, ctx: PaintContext, r: Rect, quality: string, selected?: boolean): void {
	const theme = ctx.theme;
	const radius = theme.radius.md;
	const state = ctx.state;
	const fill = state.disabled ? withAlpha(theme.colors.surface.sunken, theme.painter.disabledAlpha) : theme.colors.surface.sunken;
	let stroke = itemQualityColor(ctx, quality);
	if (state.pressed) stroke = theme.colors.text.primary;
	if (selected === true || state.selected) stroke = theme.colors.accent.primary;
	nvg.BeginPath();
	nvg.RoundedRect(r.x, r.y, r.width, r.height, radius);
	nvg.FillColor(Color(withAlpha(fill, ctx.opacity)));
	nvg.Fill();
	if (quality !== "empty" && !state.disabled) {
		nvg.BeginPath();
		nvg.RoundedRect(r.x + 3, r.y + 3, r.width - 6, r.height - 6, math.max(1, radius - 2));
		nvg.FillColor(Color(withAlpha(stroke, selected === true || state.selected ? 0.24 : 0.12)));
		nvg.Fill();
	}
	nvg.BeginPath();
	nvg.RoundedRect(r.x + 0.5, r.y + 0.5, r.width - 1, r.height - 1, radius);
	nvg.StrokeWidth(selected === true || state.selected ? theme.stroke.normal : theme.stroke.hairline);
	nvg.StrokeColor(Color(withAlpha(stroke, state.disabled ? 0.38 : ctx.opacity)));
	nvg.Stroke();
}
